### ISSUE TITLE
expandSuperTypes not using custom classloader

### ISSUE THREAD

**Issue by strmer15** [ronmamo/reflections#190]
Reflections version: 0.9.11

I created a Maven Mojo for my project that would do build-time checks for problems after compilation succeeded.  However, when I built my Reflections object and used the SubTypesScanner, I would get an error when running the plugin: 

```
org.reflections.ReflectionsException: could not get type for name <some class>
	at org.reflections.ReflectionUtils.forName(ReflectionUtils.java:390)
	at org.reflections.Reflections.expandSuperTypes(Reflections.java:381)
	at org.reflections.Reflections.<init>(Reflections.java:126)
	at <my plugin class>
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:309)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:194)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:107)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:993)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:345)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:191)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: java.lang.ClassNotFoundException: <some class>
	at org.codehaus.plexus.classworlds.strategy.SelfFirstStrategy.loadClass(SelfFirstStrategy.java:50)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.unsynchronizedLoadClass(ClassRealm.java:271)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:247)
	at org.codehaus.plexus.classworlds.realm.ClassRealm.loadClass(ClassRealm.java:239)
	at org.reflections.ReflectionUtils.forName(ReflectionUtils.java:388)
	... 25 more
```

My Reflections object looked like this:

```
final Reflections reflections = new Reflections(new ConfigurationBuilder()
                .addUrls(ClasspathHelper.forPackage("my.package", classLoader))
                .addUrls(ClasspathHelper.forPackage("my.other.package", classLoader))
                .addClassLoaders(classLoader)
                .setScanners(new MethodAnnotationsScanner(), new SubTypesScanner(), new ResourcesScanner()));
```

My classloader is a custom URLClassLoader built from the Maven project's runtime classpath elements:

```
        final URLClassLoader classLoader;
        try {
            final List<String> classpathElements = project.getRuntimeClasspathElements();

            final URL[] urls = classpathElements.stream()
                    .map(s -> {
                        try {
                            return new File(s).toURI().toURL();
                        } catch (Exception e) {
                            throw new RuntimeException("Could not convert to URL: " + s, e);
                        }
                    })
                    .collect(Collectors.toList())
                    .toArray(new URL[classpathElements.size()]);

            classLoader = new URLClassLoader(urls,
                    Thread.currentThread().getContextClassLoader());

        } catch (DependencyResolutionRequiredException e) {
            throw new MojoExecutionException("Could not load the classpath elements", e);
        }
```

It turns out that the classLoader I pass into the Reflections configuration is not being used in the expandSuperTypes method.  From the Reflections source code:

```
    public void expandSuperTypes() {
        if (store.keySet().contains(index(SubTypesScanner.class))) {
            Multimap<String, String> mmap = store.get(index(SubTypesScanner.class));
            Sets.SetView<String> keys = Sets.difference(mmap.keySet(), Sets.newHashSet(mmap.values()));
            Multimap<String, String> expand = HashMultimap.create();
            for (String key : keys) {
                final Class<?> type = forName(key);
                if (type != null) {
                    expandSupertypes(expand, key, type);
                }
            }
            mmap.putAll(expand);
        }
    }
```

The line ```final Class<?> type = forName(key);``` should be passing in the classloader, but is not.  When I created a custom class that extended Reflections and called ```final Class<?> type = forName(key, loaders());``` instead, it worked correctly.

**Comment by metasim** [comment]
@strmer15 You found a workaround for this, aside from having a custom build? 

Given the backlog of pull requests I worry about these sorts of things getting addressed. :-/

**Comment by strmer15** [comment]
@metasim I created a custom class extending the org.reflections.Reflections class, then overrode the expandSuperTypes() method.  There were three private methods I had to copy/paste from the Reflections class itself, as you can see below.

```
public class MyReflections extends Reflections {

    public MyReflections() {
    }

    public MyReflections(final Configuration configuration) {
        super(configuration);
    }

    public MyReflections(final String prefix, final @Nullable Scanner... scanners) {
        super(prefix, scanners);
    }

    public MyReflections(final Object... params) {
        super(params);
    }

    @Override
    public void expandSuperTypes() {
        if (store.keySet().contains(index(SubTypesScanner.class))) {
            Multimap<String, String> mmap = store.get(index(SubTypesScanner.class));
            Sets.SetView<String> keys = Sets.difference(mmap.keySet(), Sets.newHashSet(mmap.values()));
            Multimap<String, String> expand = HashMultimap.create();
            for (String key : keys) {
                final Class<?> type = ReflectionUtils.forName(key, loaders());
                if (type != null) {
                    expandSupertypes(expand, key, type);
                }
            }
            mmap.putAll(expand);
        }
    }

    private void expandSupertypes(Multimap<String, String> mmap, String key, Class<?> type) {
        for (Class<?> supertype : ReflectionUtils.getSuperTypes(type)) {
            if (mmap.put(supertype.getName(), key)) {
                if (Reflections.log != null) Reflections.log.debug("expanded subtype {} -> {}", supertype.getName(), key);
                expandSupertypes(mmap, supertype.getName(), supertype);
            }
        }
    }

    private static String index(Class<? extends Scanner> scannerClass) {
        return scannerClass.getSimpleName();
    }

    private ClassLoader[] loaders() {
        return configuration.getClassLoaders();
    }
}
```

**Comment by DoingLee** [comment]
when to release this bug fix solution to maven?

**Comment by ronmamo** [comment]
fixed in 0.9.12
